### PR TITLE
New show command

### PR DESCRIPTION
**Pull request by MeBrei**
! This breaks existing machines ! 

This creates a new show command `show-unbound` which does not add tabs/line breaks.

This simplifies signing requests in other languages, fe jsradicle.

### CODE REVIEW

File: src/Radicle/Internal/PrimFns.hs [hunk 1]
```diff
@@ -553,6 +553,9 @@ purePrimFns = fromList $ allDocs $
     , ( "show"
       , "Returns a string representing the argument value."
       , oneArg "show" (pure . String . renderPrettyDef))
+    , ( "show-unbound"
+      , "Returns a string representing the argument value. (No extra formatting)"
+      , oneArg "show-com" (pure . String . renderPrettyUnbounded))
```
Comment (jkarni) [inline src/Radicle/Internal/PrimFns.hs]:
```suggestion
      , oneArg "show-unbound" (pure . String . renderPrettyUnbounded))
```

File: src/Radicle/Internal/Pretty.hs [hunk 2]
```diff
@@ -164,6 +164,15 @@ renderCompactPretty = renderStrict . layoutCompact . prettyV
...
 renderPretty pg = renderStrict . layoutSmart (LayoutOptions pg) . prettyV
 
+-- | 'renderPretty', but with unbounded width.
+--
+-- Example:
+--
+-- >>> renderPrettyUnbounded (asValue (List [String "hi", String "there"]))
+-- "(\"hi\" \"there\")"
+renderPrettyUnbounded :: PrettyV v => v -> Text
+renderPrettyUnbounded = renderPretty Unbounded
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]: Can you add a test somewhere that generates arbitrary values that don't include strings with whitespace, and then check that `rendePrettyUnbound` never results in multiple consecutive spaces, tabs, or newlines?

Comment (MeBrei) [reply src/Radicle/Internal/Pretty.hs]: I added tests for `show-unbound` which inderectly tests `renderPrettyUnbounded`

File: test/spec/Radicle/Internal/Arbitrary.hs [hunk 3]
```diff
@@ -70,3 +53,33 @@ instance Arbitrary a => Arbitrary (Bindings a) where
 
 instance Arbitrary a => Arbitrary (Doc.Docd a) where
     arbitrary = Doc.Docd Nothing <$> arbitrary
+
+gen :: Gen Text -> Int -> Gen Value
+gen textGen n | n == 0 = frequency $ first pred <$> freqs
```
Comment (jkarni) [inline test/spec/Radicle/Internal/Arbitrary.hs]: Can you give a more descriptive name to this, and maybe a docstring?

File: test/spec/Radicle/Internal/Arbitrary.hs [hunk 4]
```diff
@@ -19,34 +22,14 @@ instance Arbitrary r => Arbitrary (Env r) where
...
-        prims = purePrimFns
-        isPrimop x = x `elem` Map.keys (getPrimFns prims)
-        isNum x = isJust (readMaybe (toS $ fromIdent x) :: Maybe Scientific)
-        lambdaArgs = oneof [ PosArgs <$> sizedList, VarArgs <$> arbitrary ]
+    arbitrary = sized (gen arbitrary)
+
+newtype NoSpaceValue = NoSpaceValue Value
+  deriving Show
+
+instance Arbitrary NoSpaceValue where
```
Comment (jkarni) [inline test/spec/Radicle/Internal/Arbitrary.hs]: Maybe rename this to `NoDoubleSpacesValue`?

File: test/spec/Radicle/Tests.hs [hunk 5]
```diff
@@ -561,6 +571,10 @@ test_eval =
   where
     failsWith src err    = noStack (runPureCode src) @?= Left err
     succeedsWith src val = runPureCode src @?= Right val
+    no_tabs t = not $ T.any (== '\t') t
+    no_line_breaks t = not $ T.any (== '\n') t
+    no_double_spaces t = not $ T.isInfixOf "  " t
+    no_unallowed_whitespace t = no_double_spaces t && no_line_breaks t && no_tabs t
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]: Camel case.

File: test/spec/Radicle/Tests.hs [hunk 6]
```diff
@@ -475,6 +475,16 @@ test_eval =
         runPureCode "(show (dict 'a 1))" @?= Right (String "{a 1}")
         runPureCode "(show (fn [x] x))" @?= Right (String "(fn [x] x)")
 
+    , testCase "'show-unbounds' works" $ do
+        runPureCode "(show-unbound {:b 2 :a 1})" @?= Right (String "{:a 1 :b 2}")
+        runPureCode "(show-unbound {:bar \"bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar\" :foo \"foo foo foo foo foo\"})"
+          @?= Right (String "{:bar \"bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar\" :foo \"foo foo foo foo foo\"}")
+
+    , testProperty "'show-unbounds does not add any whitespace characters'" $ \(NoSpaceValue val :: NoSpaceValue) -> do
+        let res = runPureCode $ toS [i|(show-unbound (quote #{renderPrettyUnbounded val}))|]
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]: Is the quote needed?

File: src/Radicle/Internal/PrimFns.hs [hunk 7]
```diff
@@ -553,6 +553,9 @@ purePrimFns = fromList $ allDocs $
     , ( "show"
       , "Returns a string representing the argument value."
       , oneArg "show" (pure . String . renderPrettyDef))
+    , ( "show-unbound"
+      , "Returns a string representing the argument value. (No extra formatting)"
```
Comment (jkarni) [inline src/Radicle/Internal/PrimFns.hs]: Maybe `show-unformatted`?

File: test/spec/Radicle/Tests.hs (hunk 6 above)
Comment (MeBrei) [reply test/spec/Radicle/Tests.hs]: Should only be around the function name

Comment (jkarni) [reply test/spec/Radicle/Tests.hs]: Not sure I understand.

Right now it seems like some pretty-printing is not being tested because of quote. E.g., lambdas.

Comment (MeBrei) [reply test/spec/Radicle/Tests.hs]: Sorry, misread your message (which pointed me to a different typo)

The quote is needed as otherwise on show/show-unformatted unknown identifiers (that are arbitrarily created) lead to errors. Quoting prevents that. I don't see what the problem with lambdas is, at least from testing in repl.

Comment (jkarni) [reply test/spec/Radicle/Tests.hs]:
```
rad> (type (quote (fn [] 3)))
:list
rad> (type (fn [] 3))
:function
```
So you're never giving `show-unbound` an argument that's a function, as far as I can tell.

You can just make the test assert that *if* the result is not an error, it satisfies your property (or else discard errors).
